src/libstore/aws-creds.cc

```diff
@@ -0,0 +1,219 @@
+{
+    if (!provider || !provider->IsValid()) {
+        throw AwsAuthError("AWS credential provider is invalid");
+    }
+
+    struct State
+    {
+        std::optional<AwsCredentials> result;
+        int resolvedErrorCode = 0;
+        bool resolved = false;
+    };
+
+    Sync<State> state;
+    std::condition_variable cv;
+
+    provider->GetCredentials([&](std::shared_ptr<Aws::Crt::Auth::Credentials> credentials, int errorCode) {
+        auto state_ = state.lock();
+
+        if (errorCode != 0 || !credentials) {
+            state_->resolvedErrorCode = errorCode;
+        } else {
+            auto accessKeyId = Aws::Crt::ByteCursorToStringView(credentials->GetAccessKeyId());
+            auto secretAccessKey = Aws::Crt::ByteCursorToStringView(credentials->GetSecretAccessKey());
+            auto sessionToken = Aws::Crt::ByteCursorToStringView(credentials->GetSessionToken());
+
+            std::optional<std::string> sessionTokenStr;
+            if (!sessionToken.empty()) {
+                sessionTokenStr = std::string(sessionToken.data(), sessionToken.size());
+            }
+
+            state_->result = AwsCredentials(
+                std::string(accessKeyId.data(), accessKeyId.size()),
+                std::string(secretAccessKey.data(), secretAccessKey.size()),
+                sessionTokenStr);
+        }
+
+        state_->resolved = true;
+        cv.notify_one();
+    });
+
+    {
+        auto state_ = state.lock();
+        // AWS CRT GetCredentials is asynchronous and only guarantees the callback will be
+        // invoked if the initial call returns success. There's no documented timeout mechanism,
+        // so we add a timeout to prevent indefinite hanging if the callback is never called.
+        // Use an absolute deadline to handle spurious wakeups correctly.
+        auto timeout = std::chrono::seconds(30);
+        auto deadline = std::chrono::steady_clock::now() + timeout;
+
+        while (!state_->resolved) {
+            if (state_.wait_until(cv, deadline) == std::cv_status::timeout) {
+                // Double-check the condition after timeout to avoid race
+                if (!state_->resolved) {
+                    throw AwsAuthError(
+                        "Timeout waiting for AWS credentials (%d seconds)",
+                        std::chrono::duration_cast<std::chrono::seconds>(timeout).count());
+                }
+                break;
+            }
+        }
+    }
+
+    auto state_ = state.lock();
+    if (!state_->result) {
+        throw AwsAuthError("Failed to resolve AWS credentials: error code %d", state_->resolvedErrorCode);
+    }
+
+    return *state_->result;
+}
+
+// Global credential provider cache using boost's concurrent map
+// Key: profile name (empty string for default profile)
+using CredentialProviderCache =
+    boost::concurrent_flat_map<std::string, std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider>>;
+
+static CredentialProviderCache credentialProviderCache;
```

Comment:
The credentials have a expiration https://awslabs.github.io/aws-crt-cpp/class_aws_1_1_crt_1_1_auth_1_1_credentials.html

You might want to use a different value that is the pointer + expiration and check it.

Reply:
The credentials have an expiration, but the providers don't, right?

Reply:
> I think the providers already cache themselves but not 100% sure (was reading c++ SDK)

I also couldn't be sure, so I played it safe. If you can show that they stay auto-cached in bootstrap I'd love to get rid of some of this